### ISSUE TITLE
Alignment score

### ISSUE THREAD

**Issue by dgm2** [PredictiveIntelligenceLab/jaxpi#30]
for the (intra) alignment score, would something like this be ok?
because its bouncing between positive and negative.
if you have a snippet, could you share it? or some draft.
Best regards

```python
def alignment_score_jax3(vectors):
    vectors = jnp.vstack(vectors)
    norms = jnp.linalg.norm(vectors, axis=1, keepdims=True)
    unit_vectors = vectors / norms
    mean_unit_vector = jnp.mean(unit_vectors, axis=0)
    mean_norm_squared = jnp.sum(mean_unit_vector ** 2)
    score = 2 * mean_norm_squared - 1
    return score

def loss_grad(params, weights, batch, loss_name):
    def single_loss(params):
        losses = self.weighted_losses(params, weights, batch, *args)
        return losses[loss_name]

    return grad(single_loss)(params)

multi_grads = {name: loss_grad(state.params, state.weights, batch, name) for name in names}
from jax.flatten_util import ravel_pytree
vecs=[]
for m in multi_grads.items():
    (k,p)=m
    v, _ = ravel_pytree(p)
    vecs.append(v)
res=alignment_score_jax3(vecs)
```

**Comment by sifanexisted** [comment]
Please note that we compute gradient alignment scores after applying the optimizer's gradient transformation. You may use the following as a reference.


```
def compute_intra_alignment_score(a, b):
    a = flatten_pytree(a)
    b = flatten_pytree(b)

    score = a / jnp.linalg.norm(a) + b / jnp.linalg.norm(b)
    score = jnp.linalg.norm(score) / 2
    return 2 * score**2 - 1

def compute_inter_alignment_score(state, grad_dict, epsilon = 1e-8):
    score = 0.0
    scaled_score = 0.0
    for key, value in grad_dict.items():
        scaled_value, _ = state.tx.update(value, state.opt_state, state.params)

        # Flatten the pytrees into 1D vectors
        value = flatten_pytree(value)
        scaled_value = flatten_pytree(scaled_value)

        score += value / (jnp.linalg.norm(value) + epsilon)
        scaled_score += scaled_value / (jnp.linalg.norm(scaled_value) + epsilon)

    score = jnp.linalg.norm(score) / len(grad_dict)
    scaled_score = jnp.linalg.norm(scaled_score) / len(grad_dict)

    score = 2 * score**2 - 1
    scaled_score = 2 * scaled_score**2 - 1

    return score, scaled_score

  grads = grad(self.model.loss)(state.params, state.weights, batch)
  grads_dict = jacrev(self.model.losses)(state.params, batch)

  # optimizer update
  scaled_grads, _ = state.tx.update(grads, state.opt_state, state.params)

  old_grads = grad(self.model.loss)(self.model.old_state.params, self.model.old_state.weights, batch)
  scaled_old_grads, _ = state.tx.update(old_grads, self.model.old_state.opt_state,
                                        self.model.old_state.params)

  inter_align, inter_scaled_align = compute_inter_alignment_score(state, grads_dict)

  intra_align = compute_intra_alignment_score(grads, old_grads)
  intra_scaled_align = compute_intra_alignment_score(scaled_grads, scaled_old_grads)
```
